### PR TITLE
feat: allow "R source package with Rust library binary" installation

### LINKED ISSUE #408: `arrow` like "R source package with Rust library binary" installation
The arrow package will attempt to download a pre-built binary from the Internet if `NOT_CRAN=true`.
<https://arrow.apache.org/docs/11.0/r/articles/install.html#r-source-package-with-libarrow-binary>

This method also has the advantage that many features disabled by default (i.e., on CRAN) during source installation are enabled in the pre-built binaries.

I thought the same thing could be done here and have recently added that functionality to `prqlr` (eitsupi/prqlr#195).
The arm64 architecture can also be supported due to cargo's excellent cross-compilation, and the glibc version should be negligible by selecting the musl target. (e.g. #86)

Unlike the binary installation via GitHub releases that this repository currently offers, it has the advantage of being valid for installation from anywhere. Like:

```r
Sys.setenv(NOT_CRAN="true")
remotes::install_github("eitsupi/prqlr")
```

```r
Sys.setenv(NOT_CRAN="true")
install.packages("prqlr", repos = "https://eitsupi.r-universe.dev/") # Of course, Windows and macOS do binary installations
```

For example, the installation on arm64 Linux from R-universe is as follows:

```log
$ docker run --rm -it rocker/r-ver bash
root@3e5a9638fea4:/# R

R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: aarch64-unknown-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> Sys.setenv(NOT_CRAN="true")
> install.packages("prqlr", repos = "https://eitsupi.r-universe.dev/")
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://eitsupi.r-universe.dev/src/contrib/prqlr_0.5.3.9000.tar.gz'
Content type 'application/x-gzip' length 191638 bytes (187 KB)
==================================================
downloaded 187 KB

* installing *source* package ‘prqlr’ ...
** using staged installation

--------------- [ TRY TO DOWNLOAD PRE-BUILT BINARY ] ---------------
Found pre-built binary at <https://github.com/eitsupi/prqlr/releases/download/lib-v0.9.0/libprqlr-0.9.0-aarch64-unknown-linux-musl.tar.gz>.
Downloading...
Checking SHA256 for </tmp/RtmpQpUMZg/filed6477f9eb5.tar.gz>...
SHA256 matches for </tmp/RtmpQpUMZg/filed6477f9eb5.tar.gz>.
Extracted pre-built binary to </tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/tools> directory.
--------------------------------------------------------------------


---------------------- [LIBRARY BINARY FOUND] ----------------------
The library was found at <tools/libprqlr.a>. No need to build it.
--------------------------------------------------------------------

** libs
using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
rm -Rf "prqlr.so" "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target//release/libprqlr.a" "entrypoint.o"
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include    -fPIC  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c entrypoint.c -o entrypoint.o
if [ -f "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/../tools/libprqlr.a" ]; then \
        mkdir -p "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target//release" ; \
        mv "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/../tools/libprqlr.a" "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target//release/libprqlr.a" ; \
        exit 0; \
fi && \
if [ -f "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor.tar.xz" ]; then \
        mkdir -p "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor" && \
        /usr/bin/tar --extract --xz --file "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor.tar.xz" -C "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor" && \
        mkdir -p "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/.cargo" && \
        cp "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor-config.toml" "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/.cargo/config.toml"; \
fi && \
if [ "true" != "true" ]; then \
        export CARGO_HOME="/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/.cargo"; \
        export CARGO_BUILD_JOBS=2; \
fi && \
        export PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/.cargo/bin" && \
        cargo build --lib --manifest-path="/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/Cargo.toml" --target-dir "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target" --target="" \
                --profile="release" --features=""
if [ "true" != "true" ]; then \
        rm -Rf "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/.cargo" "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/vendor" "/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target//release/build"; \
fi
gcc -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o prqlr.so entrypoint.o -L/tmp/RtmpgWnDdc/R.INSTALL1b54f3616/prqlr/src/rust/target//release -lprqlr -L/usr/local/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-prqlr/00new/prqlr/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (prqlr)

The downloaded source packages are in
        ‘/tmp/RtmpohJEuV/downloaded_packages’
```

I would like to bring this same thing here. What do you think?
The administrative disadvantage is the need for additional library versioning and release work.
(Don't forget to raise the version on Cargo.toml when making changes on the Rust side.)

### PR DESCRIPTION
**Pull request by eitsupi**
Close #408

Allows us to download binaries from GitHub Releases and skip the `cargo build` step using the mechanism copied from `prqlr`.

Like this:

```r
> Sys.setenv(NOT_CRAN="true")
> remotes::install_github("pola-rs/r-polars#450")
Downloading GitHub repo pola-rs/r-polars@bin-lib-install
Running `R CMD build`...
* checking for file ‘/tmp/Rtmpghk68j/remotes3e78001b4b/pola-rs-r-polars-9ece233/DESCRIPTION’ ... OK
* preparing ‘polars’:
* checking DESCRIPTION meta-information ... OK
* cleaning src
* running ‘cleanup’
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘polars_0.8.1.9000.tar.gz’
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
* installing *source* package ‘polars’ ...
** using staged installation

--------------- [ TRY TO DOWNLOAD PRE-BUILT BINARY ] ---------------
Found pre-built binary at <https://github.com/pola-rs/r-polars/releases/download/lib-v0.33.0/libr_polars-0.33.0-x86_64-unknown-linux-gnu.tar.gz>.
Downloading...
Checking SHA256 for </tmp/RtmpcX8QCj/file17b333984b8.tar.gz>...
SHA256 matches for </tmp/RtmpcX8QCj/file17b333984b8.tar.gz>.
Extracted pre-built binary to </tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/tools> directory.
--------------------------------------------------------------------


---------------------- [LIBRARY BINARY FOUND] ----------------------
The library was found at </tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/tools/libr_polars.a>. No need to build it.
--------------------------------------------------------------------

** libs
using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
rm -Rf polars.so /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a entrypoint.o
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include    -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c entrypoint.c -o entrypoint.o
if [ -f "/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/../tools/libr_polars.a" ]; then \
        mkdir -p "/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release" ; \
        mv "/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/../tools/libr_polars.a" "/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a" ; \
        exit 0; \
fi && \
if [ "true" != "true" ]; then \
        export CARGO_HOME=/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/.cargo; \
fi && \
export PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/.cargo/bin" && \
if [ "" = "true" ]; then \
        cargo build --target= --lib --profile release --manifest-path="./rust/Cargo.toml" --features "full_features"; \
else \
        cargo build --target= --lib --profile release --manifest-path="./rust/Cargo.toml"; \
fi
if [ "true" != "true" ]; then \
        rm -Rf /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/.cargo && \
        rm -Rf /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/build; \
fi
if [ -f "/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a" ]; then \
        echo "file is there: "; \
elif [ -f "./rust/target//release/libr_polars.a" ]; then \
        echo "file is './rust/target//release/libr_polars.a'"; \
        mkdir -p /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release ; \
        echo "trying to symlink in ./rust/target//release/libr_polars.a"; \
        ln -s ./rust/target//release/libr_polars.a /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a ; \
fi
file is there:
if [ "" = "true" ]; then \
        echo "cleanup!!" ; \
        mv /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/../temp_binary.a; \
        rm -rf /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release; \
        mkdir /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release; \
        mv /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/../temp_binary.a /tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release/libr_polars.a; \
        rm -rf ./src/.cargo; \
else \
        echo "hands off!!" ; \
fi
hands off!!
gcc -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o polars.so entrypoint.o -L/tmp/RtmpGSDwc8/R.INSTALLb33d8fe552/polars/src/rust/target//release -lr_polars -L/usr/local/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-polars/00new/polars/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (polars)
```

### CODE REVIEW

**Comment by etiennebacher** [comment]
I must admit I'm a bit lost on the changes in CI and in the procedure to build/install `polars` so it would be great to have a summary of the changes somewhere in the docs when all of this is done :smile: (it doesn't have to be in this PR though)

**Comment by eitsupi** [comment]
> I must admit I'm a bit lost on the changes in CI and in the procedure to build/install `polars` so it would be great to have a summary of the changes somewhere in the docs when all of this is done 😄 (it doesn't have to be in this PR though)

Definitely. I was imagining something like `DEVELOPMENT.md`

**Comment by eitsupi** [comment]
After merging this, builds on R-universe should use the pre-built binaries and SIMD should be enabled even when installing from R-universe.
I would like to update the documentation after checking it.

File: configure [hunk 1]
```diff
@@ -39,6 +39,13 @@ check_bin_lib() {
     LIBR_POLARS_BUILD="false"
   fi
 
+  # On R-universe, we try to download the pre-built binary from GitHub releases.
+  if [ -z "${NOT_CRAN}" ] && [ -n "${MY_UNIVERSE}" ]; then
+    echo "It seems that this is on R-universe <${MY_UNIVERSE}>."
+    echo "Trying to download pre-built binary."
+    LIBR_POLARS_BUILD="false"
+  fi
```
Comment (eitsupi) [inline configure:47]: This step detects R-universe and enable install with pre-built binary.

**Review by etiennebacher — APPROVED** [review]
I've no clue about this, so just merge whenever you think it's ready

**Pull request merged**